File: examples/ts_tesseract/ocrTesseract.ts
/**
 * Reference OCR adapter: Tesseract.js -> OcrLine[].
 *
 * OPTIONAL — not part of the parser core. Bring any OCR you like; this just shows
 * one way to produce the `OcrLine`s the parser consumes.
 *
 *   npm install tesseract.js image-size costco-price-tag-parser
 *
 * Note: tesseract.js's result shape varies across major versions. This targets
 * v5 (`recognize(img, {}, { blocks: true })` -> data.blocks[].paragraphs[].lines[]).
 * Adjust the traversal if your installed version differs.
 */

import { readFileSync } from "node:fs";

import type { OcrLine } from "costco-price-tag-parser";
import imageSize from "image-size";
import { createWorker } from "tesseract.js";

export async function recognize(imagePath: string, lang = "kor+eng"): Promise<OcrLine[]> {
  const dim = imageSize(readFileSync(imagePath));
  const width = dim.width ?? 1;
  const height = dim.height ?? 1;

  const worker = await createWorker(lang);
  try {
    const { data } = await worker.recognize(imagePath, {}, { blocks: true });
    const out: OcrLine[] = [];
    for (const block of data.blocks ?? []) {
      for (const para of block.paragraphs ?? []) {
        for (const ln of para.lines ?? []) {
          const text = ln.text.trim();
          if (!text) continue;
          const b = ln.bbox; // pixels: { x0, y0, x1, y1 }
          out.push({
            text,
            x: b.x0 / width,
            yTop: b.y0 / height,
            w: (b.x1 - b.x0) / width,
            h: (b.y1 - b.y0) / height,
            conf: Math.max(0, ln.confidence / 100),
          });
        }
      }
    }
    // Reading order: top-to-bottom, then left-to-right (what the parser expects).
    out.sort((a, b) => (Math.abs(a.yTop - b.yTop) > 0.02 ? a.yTop - b.yTop : a.x - b.x));
    return out;
  } finally {
    await worker.terminate();
  }
}

File: examples/python_tesseract/ocr_tesseract.py
"""Reference OCR adapter: Tesseract -> list[OcrLine].

OPTIONAL — not part of the parser core. Bring any OCR you like; this just shows
one way to produce the `OcrLine`s the parser consumes.

Requires the tesseract binary (with Korean traineddata) and:
    pip install pytesseract Pillow
"""

from __future__ import annotations

import pytesseract
from PIL import Image

from costco_price_tag_parser import OcrLine


def recognize(image_path: str, lang: str = "kor+eng") -> list[OcrLine]:
    img = Image.open(image_path)
    width, height = img.size
    data = pytesseract.image_to_data(img, lang=lang, output_type=pytesseract.Output.DICT)

    # Group word boxes into text lines by (block, paragraph, line).
    groups: dict[tuple[int, int, int], dict] = {}
    for i in range(len(data["text"])):
        text = data["text"][i].strip()
        if not text:
            continue
        key = (data["block_num"][i], data["par_num"][i], data["line_num"][i])
        left, top = data["left"][i], data["top"][i]
        right, bottom = left + data["width"][i], top + data["height"][i]
        g = groups.setdefault(
            key, {"words": [], "left": left, "top": top, "right": right, "bottom": bottom, "conf": []}
        )
        g["words"].append(text)
        g["left"] = min(g["left"], left)
        g["top"] = min(g["top"], top)
        g["right"] = max(g["right"], right)
        g["bottom"] = max(g["bottom"], bottom)
        g["conf"].append(float(data["conf"][i]))

    lines: list[OcrLine] = []
    for g in groups.values():
        lines.append(
            OcrLine(
                text=" ".join(g["words"]),
                x=g["left"] / width,
                y_top=g["top"] / height,
                w=(g["right"] - g["left"]) / width,
                h=(g["bottom"] - g["top"]) / height,
                conf=max(0.0, sum(g["conf"]) / len(g["conf"]) / 100.0),
            )
        )

    # Reading order: top-to-bottom, then left-to-right (what the parser expects).
    lines.sort(key=lambda o: (round(o.y_top, 2), o.x))
    return lines
